### PR TITLE
evaluate_horner for mpoly, non-recursive

### PR DESCRIPTION
**Pull request by tthsqe12**

### CODE REVIEW

File: src/algorithms/MPolyEvaluate.jl [hunk 1]
```diff
@@ -0,0 +1,289 @@
...
+    return ctxC(Bcoeffs[1+rtypes[1+rp]])
+  end
+end
+
+function evaluate_horner(B::MPolyElem, C::Vector{<:RingElem})
+  @assert nvars(parent(B)) <= length(C)
+  ctxC = parent(C[1])
+  Bcoeffs = collect(coefficients(B))
+  ctxA = parent(one(coefficient_ring(B)) * one(ctxC))
+  ctxA === ctxC || error("probably not implemented")
```
Comment (thofma) [inline src/algorithms/MPolyEvaluate.jl:285]: So the code cannot handle fmpq_mpoly evaluating at fmpz's?

File: src/algorithms/MPolyEvaluate.jl [hunk 2]
```diff
@@ -0,0 +1,289 @@
...
+
+function mul!(z::T, a, b)::T where T
+  return a*b
+end
+
+function pow!(z::T, a, b)::T where T
+  return a^b
+end
+
+# a *= x^p, using t as temp. Must return the modifed a and t
```
Comment (thofma) [inline src/algorithms/MPolyEvaluate.jl:19]: Must `t == x^p` hold?

Comment (tthsqe12) [reply src/algorithms/MPolyEvaluate.jl:19]: of course not

File: src/algorithms/MPolyEvaluate.jl (hunk 1 above)
Comment (tthsqe12) [reply src/algorithms/MPolyEvaluate.jl:285]: Currently cannot. Will try.

Comment (thofma) [reply src/algorithms/MPolyEvaluate.jl:285]: I don't quite understand. You have the generic fallback, so why is it not working?

File: src/algorithms/MPolyEvaluate.jl (hunk 2 above)
Comment (thofma) [reply src/algorithms/MPolyEvaluate.jl:19]: Why must it be returned then? I don't quite understand. Can you elaborate?

File: src/algorithms/MPolyEvaluate.jl (hunk 1 above)
Comment (tthsqe12) [reply src/algorithms/MPolyEvaluate.jl:285]: you have to predict the type of the return of a+b. It will be working soon, but requres some changes.

**Comment by tthsqe12** [comment]
This is discrimination: codecov considers labels (targets of gotos) as unconvered.

File: src/algorithms/MPolyEvaluate.jl (hunk 1 above)
Comment (thofma) [reply src/algorithms/MPolyEvaluate.jl:285]: ok, thanks

Comment (tthsqe12) [reply src/algorithms/MPolyEvaluate.jl:285]: do we have any example where a+b lies in neither parent(a) nor parent(b)?

**Comment by tthsqe12** [comment]
@thofma It now should work with either combination of promotions, with the actual promotions hidded in the arithmetic operations.

**Comment by thofma** [comment]
Very nice, thanks.

File: src/algorithms/MPolyEvaluate.jl (hunk 1 above)
Comment (thofma) [reply src/algorithms/MPolyEvaluate.jl]: Sorry, I missed your message. Yes, I think this is possible.

**Pull request merged**